docs/gui/menus.md

```diff
@@ -0,0 +1,343 @@
+  // ...
+}
+
+@Override
+public boolean stillValid(Player player) {
+  return this.stillValid.apply(player);
+}
+```
+
+### `ContainerLevelAccess`
+
+If a menu is opened by a block, sometimes the menu needs access to its location in a given level. This can be done by passing in a `ContainerLevelAccess` to the menu. The access supplies the current level and location of the block within an enclosed scope. When constructing the menu on the server, a new access can be created by calling `ContainerLevelAccess#create`. The client menu constructor can pass in `ContainerLevelAccess#NULL`, which will do nothing.
+
+```java
+// Client menu constructor
+public MyMenuAccess(int containerId, Inventory playerInventory) {
+  this(containerId, playerInventory, ContainerLevelAccess.NULL);
+}
+
+// Server menu constructor
+public MyMenuAccess(int containerId, Inventory playerInventory, ContainerLevelAccess access) {
+  // ...
+}
+```
+
+### Data Synchronization
+
+Some data needs to be present on both the server and the client to display to the player. To do this, the menu implements a basic layer of data synchronization such that whenever the current data does not match the data last synced to the client. For players, this is checked every tick.
+
+Minecraft supports two forms of data synchronization by default: `ItemStack`s via `Slot`s and integers via `DataSlot`s. `Slot`s and `DataSlot`s are views which hold references to data storages that can be be modified by the player in a screen, assuming the action is valid. These can be added to a menu within the constructor through `#addSlot` and `#addDataSlot`.
+
+!!! note
+    Since `Container`s used by `Slot`s are deprecated by Forge in favor of using the [`IItemHandler` capability][cap], the rest of the explanation will revolve around using `SlotItemHandler`: the capability variant.
+
+A `SlotItemHandler` contains four parameters: the `IItemHandler` representing the inventory the stacks are within, the index of the stack this slot is specifically representing, and the x and y position of where the top-left position of the slot will render on the screen relative to `AbstractContainerScreen#leftPos` and `#topPos`. The client menu constructor should always supply an empty instance of an inventory of the same size.
+
+In most cases, any slots the menu contains is first added, followed by the player's inventory, and finally concluded with the player's hotbar. To access any individual `Slot` from the menu, the index must be calculated based upon the order of which slots were added.
+
+A `DataSlot` is an abstract class which should implement a getter and setter to reference the data stored in the data storage object.The client menu constructor should always supply a new instance via `DataSlot#standalone`.
+
+These, along with slots, should be recreated every time a new menu is initialized.
+
+!!! warning
+    Although a `DataSlot` may hold enough data for an integer, only a short is sent across the network.
```

Comment:
Rephrased it a bit with a bit more emphasis to draw the eye. Also, what happens when a value larger than a short is provided in a DataSlot? Is it truncated to fit the bits of a short, or can some form of overflow/underflow occur?

```suggestion
!!! warning
    Although a `DataSlot` stores an integer, it is effectively limited to a **short** (-32768 to 32767) because of how it sends the value across the network.
```